plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CombinedIcebergSplit.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.iceberg;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.MoreObjects.ToStringHelper;
+import com.google.common.collect.ImmutableList;
+import io.trino.spi.HostAddress;
+import io.trino.spi.connector.ConnectorSplit;
+
+import java.util.Collection;
+import java.util.List;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class CombinedIcebergSplit
```

Comment:
Did you try assigning split weight (as in hive) instead?

Reply:
How do weighted split work?
does it allow splits to be combined to avoid scheduling overhead?

Reply:
> How do weighted split work?

Check https://github.com/trinodb/trino/pull/9059. It's pretty well described there.

> does it allow splits to be combined to avoid scheduling overhead?

Scheduling overhead itself is minimal. Split queues might be too short if there is a lot of splits, but that's what weighted split scheduling solves

Reply:
I like that provided this solves the problem.

@alexjo2144 can you please make another version of the improvement, based on weighted splits?

Reply:
Here's combined splits + split weights:
```
trino> select count(*) from iceberg.iceberg_tpch_sf1000_orc.lineitem_small_files;
   _col0    
------------
 5999989709 
(1 row)

Query 20220519_200949_00011_empm9, FINISHED, 6 nodes
Splits: 7,469 total, 7,469 done (100.00%)
24.89 [6B rows, 7.25TB] [241M rows/s, 298GB/s]
```

Reply:
> So still faster than we have on master, but there seems to still be some extra overhead from the number of splits being so high.

How do you assign split weight? Is it similar to Hive connector approach?

> but there seems to still be some extra overhead from the number of splits being so high.

That is an assumption. It can either be:
* some kind of overhead
* workers still get starved as queue size is not sufficient even with weighted splits. You can actually try increasing split queue size (`node-scheduler.max-splits-per-node`, `node-scheduler.max-pending-splits-per-task`) to see if it helps. Is Iceberg generating splits fast enough? Check `query.schedule-split-batch-size`.

I would like to learn where the overhead comes from since we already have a mechanism (weighted splits) that should address the issue of multiple small splits. It would be better to fix/tune it rather than add yet another mechanism.

If the conclusion is that the overhead is X and it can be only addressed in Iceberg with merging splits, that's fine. However, if we start packing splits together for other connectors we should know where is the overhead.

Reply:
> I would like to learn where the overhead comes from since we already have a mechanism (weighted splits) that should address the issue of multiple small splits. It would be better to fix/tune it rather than add yet another mechanism.

That's a good point

> If the conclusion is that the overhead is X and it can be only addressed in Iceberg with merging splits, that's fine.

or we can even think of making engine-side equivalent of these changes, so connectors don't have to do this

Reply:
> or we can even think of making engine-side equivalent of these changes, so connectors don't have to do this

Sure, but then you probably don't need weighted split scheduling. Also, large splits are not neccecerly good because they provide less granularity of work which leads to work skewness (especially when there are more nodes)